src/function/gds/shortest_paths.cpp

```diff
@@ -15,57 +19,169 @@ using namespace kuzu::graph;
 namespace kuzu {
 namespace function {
 
-struct ShortestPathBindData final : public GDSBindData {
+/**
+ * A GDSFrontier implementation that keeps the lengths of the shortest paths from the source node to
+ * destination nodes. This is a light-weight implementation that can keep lengths up to and
+ * including 254. The length stored for the source node is 0. Length 255 is reserved for marking
+ * nodes that are not visited. This class is intended to represent both the current and next
+ * frontiers. At iteration i of the shortest path algorithm (iterations start from 0), nodes
+ * with mask values i are in the current frontier. Nodes with any other values are not in the
+ * frontier. Similarly at iteration i setting a node u active sets its mask value to i+1.
+ * Therefore this class needs to be informed about the current iteration of the algorithm to
+ * provide correct implementations of isActive and setActive functions.
+ */
+class PathLengths : public GDSFrontier {
```

Comment:
I understand why u are doing this and this kind of explains why you do not have two GDSFrontier within Frontiers. Though I think you can still try to put this logic in PathLengthsFrontiers? e.g.

```
PathLengthsFrontiers {

std::unique<PathLengths> frontier;
PathLengths* currFrontier;
PathLengths* nextFrontier' 
}
```

Both `currFrontier` and `nextFrontier` points to `frontier.`

Conceptually this will read better because we normally don't expect GDSFrontier two have the notion of current and next.

Reply:
I don't think I can do this. GDSTask logic calls setActive functions on the Frontier. So PathLengths, which is a frontier needs to know the currentIteration number because setting a node active depends (for this optimization) on the current iteration number.